File: packages/usage/go/cursor.go
package main

import (
	"database/sql"
	"encoding/json"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	_ "modernc.org/sqlite"
)

func parseCursor(home string, emit func(usageEvent)) (parserStats, bool, error) {
	path := filepath.Join(home, "Library", "Application Support", "Cursor", "User", "globalStorage", "state.vscdb")
	info, err := os.Stat(path)
	if err != nil {
		if os.IsNotExist(err) {
			return parserStats{}, false, nil
		}
		return parserStats{}, false, err
	}
	started := time.Now()

	dsn := (&url.URL{
		Scheme:   "file",
		Path:     path,
		RawQuery: "immutable=1&mode=ro",
	}).String()
	db, err := sql.Open("sqlite", dsn)
	if err != nil {
		return parserStats{}, false, err
	}
	defer db.Close()

	var table string
	if err := db.QueryRow(`SELECT name FROM sqlite_master WHERE type = 'table' AND name = 'cursorDiskKV' LIMIT 1`).Scan(&table); err != nil {
		if err == sql.ErrNoRows {
			return parserStats{}, false, nil
		}
		return parserStats{}, false, err
	}

	composerRows, err := loadCursorKV(db, "composerData:")
	if err != nil {
		return parserStats{}, false, err
	}
	composers := map[string]map[string]any{}
	for _, row := range composerRows {
		id := strings.TrimPrefix(row.key, "composerData:")
		if id == "" {
			continue
		}
		var composerJSON map[string]any
		if json.Unmarshal(row.value, &composerJSON) != nil {
			continue
		}
		composers[id] = composerJSON
	}

	rows, err := db.Query(`SELECT key, value FROM cursorDiskKV WHERE key LIKE 'bubbleId:%' ESCAPE '\' ORDER BY key`)
	if err != nil {
		return parserStats{}, false, err
	}
	defer rows.Close()

	seen := map[string]struct{}{}
	var buckets tokenBuckets
	events := 0
	add := func(mapped []cursorEvent) {
		for _, event := range mapped {
			if _, ok := seen[event.id]; ok {
				continue
			}
			seen[event.id] = struct{}{}
			events++
			buckets.add(event.tokens)
			emit(event.usageEvent)
		}
	}

	var currentID string
	var batch []kvRow
	flush := func() {
		if currentID == "" {
			return
		}
		if composerJSON, ok := composers[currentID]; ok {
			add(mapCursorComposer(currentID, composerJSON, batch))
			delete(composers, currentID)
		}
		batch = batch[:0]
	}
	for rows.Next() {
		var row kvRow
		if err := rows.Scan(&row.key, &row.value); err != nil {
			return parserStats{}, false, err
		}
		rest := strings.TrimPrefix(row.key, "bubbleId:")
		id, _, ok := strings.Cut(rest, ":")
		if !ok || id == "" {
			continue
		}
		if id != currentID {
			flush()
			currentID = id
		}
		batch = append(batch, row)
	}
	if err := rows.Err(); err != nil {
		return parserStats{}, false, err
	}
	flush()
	for id, composerJSON := range composers {
		add(mapCursorComposer(id, composerJSON, nil))
	}

	return finish("cursor", started, 1, info.Size(), events, buckets), true, nil
}

type cursorEvent struct {
	id string
	usageEvent
}

type kvRow struct {
	key   string
	value []byte
}

func loadCursorKV(db *sql.DB, prefix string) ([]kvRow, error) {
	escaped := strings.NewReplacer(`\`, `\\`, `%`, `\%`, `_`, `\_`).Replace(prefix)
	rows, err := db.Query(`SELECT key, value FROM cursorDiskKV WHERE key LIKE ? ESCAPE '\'`, escaped+"%")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []kvRow
	for rows.Next() {
		var row kvRow
		if err := rows.Scan(&row.key, &row.value); err != nil {
			return nil, err
		}
		out = append(out, row)
	}
	return out, rows.Err()
}

func mapCursorComposer(composerID string, composerJSON map[string]any, bubbles []kvRow) []cursorEvent {
	var bubbleEvents []cursorEvent
	for _, bubble := range bubbles {
		var bubbleJSON map[string]any
		if json.Unmarshal(bubble.value, &bubbleJSON) != nil {
			continue
		}
		tokenCount, _ := bubbleJSON["tokenCount"].(map[string]any)
		if tokenCount == nil {
			tokenCount, _ = bubbleJSON["tokenCounts"].(map[string]any)
		}
		if tokenCount == nil {
			continue
		}
		input := jsonInt(firstAny(tokenCount["inputTokens"], tokenCount["input"]))
		output := jsonInt(firstAny(tokenCount["outputTokens"], tokenCount["output"]))
		if input == 0 && output == 0 {
			continue
		}
		bubbleID := jsonString(bubbleJSON["bubbleId"])
		if bubbleID == "" {
			bubbleID = strings.TrimPrefix(bubble.key, "bubbleId:"+composerID+":")
		}
		ts := jsonTime(bubbleJSON["createdAt"])
		if ts.IsZero() {
			ts = jsonTime(composerJSON["createdAt"])
		}
		if ts.IsZero() {
			continue
		}
		model := cursorModelName(composerJSON, bubbleJSON)
		bubbleEvents = append(bubbleEvents, cursorEvent{
			id: "cursor-bubble-" + composerID + "-" + bubbleID,
			usageEvent: usageEvent{
				ts:       ts,
				agent:    "cursor",
				provider: cursorProvider(model),
				model:    model,
				tokens:   tokenBuckets{input: input, output: output},
			},
		})
	}
	// No fallback to contextTokensUsed / promptTokenBreakdown: those measure how
	// full the context window is, not tokens consumed.
	return bubbleEvents
}

func cursorModelName(composerJSON, bubbleJSON map[string]any) string {
	if bubbleJSON != nil {
		if info, ok := bubbleJSON["modelInfo"].(map[string]any); ok {
			if name := jsonString(info["modelName"]); name != "" && name != "default" {
				return name
			}
		}
	}
	if config, ok := composerJSON["modelConfig"].(map[string]any); ok {
		if selected, ok := config["selectedModels"].([]any); ok && len(selected) > 0 {
			if first, ok := selected[0].(map[string]any); ok {
				if id := jsonString(first["modelId"]); id != "" && id != "default" {
					return id
				}
			}
		}
		if name := jsonString(config["modelName"]); name != "" && name != "default" {
			return name
		}
	}
	return "cursor-auto"
}

func cursorProvider(model string) string {
	if strings.Contains(strings.ToLower(model), "grok") {
		return "xai"
	}
	return "cursor"
}

func jsonInt(value any) int64 {
	switch typed := value.(type) {
	case float64:
		return int64(typed)
	case json.Number:
		n, _ := typed.Int64()
		return n
	case int:
		return int64(typed)
	case int64:
		return typed
	case string:
		var n float64
		if json.Unmarshal([]byte(typed), &n) == nil {
			return int64(n)
		}
	}
	return 0
}

func jsonString(value any) string {
	s, _ := value.(string)
	return s
}

func jsonTime(value any) time.Time {
	switch typed := value.(type) {
	case float64:
		if typed > 1e12 {
			return time.UnixMilli(int64(typed))
		}
		if typed > 1e9 {
			return time.Unix(int64(typed), 0)
		}
	case json.Number:
		n, _ := typed.Float64()
		return jsonTime(n)
	case string:
		if ts, ok := parseTimestamp(typed); ok {
			return ts
		}
	}
	return time.Time{}
}

func firstAny(values ...any) any {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}

File: packages/usage/rust/src/cursor.rs
use crate::collect::{Parsed, Tokens, UsageEvent, finish};
use crate::parse::{failed, parse_timestamp, value_bytes};
use chrono::{DateTime, TimeZone, Utc};
use rusqlite::{Connection, OpenFlags};
use serde::Deserialize;
use serde_json::{Map, Value};
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::time::Instant;

pub fn parse_cursor(home: &Path, emit: &mut dyn FnMut(UsageEvent)) -> Parsed {
    let path = home.join("Library/Application Support/Cursor/User/globalStorage/state.vscdb");
    let meta = match std::fs::metadata(&path) {
        Ok(meta) => meta,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Parsed::default(),
        Err(error) => return failed(error),
    };
    let started = Instant::now();
    let uri = format!(
        "file:{}?immutable=1&mode=ro",
        path.to_string_lossy()
            .replace('%', "%25")
            .replace(' ', "%20")
            .replace('?', "%3f")
            .replace('#', "%23")
    );
    let mut run = || -> rusqlite::Result<Option<(usize, Tokens)>> {
        let conn = Connection::open_with_flags(
            &uri,
            OpenFlags::SQLITE_OPEN_READ_ONLY | OpenFlags::SQLITE_OPEN_URI,
        )?;
        let has_table = conn.query_row(
            "SELECT name FROM sqlite_master WHERE type = 'table' AND name = 'cursorDiskKV' LIMIT 1",
            [],
            |row| row.get::<_, String>(0),
        );
        match has_table {
            Ok(_) => {}
            Err(rusqlite::Error::QueryReturnedNoRows) => return Ok(None),
            Err(error) => return Err(error),
        }

        let mut composers: HashMap<String, CursorComposer> = HashMap::new();
        {
            let mut stmt = conn.prepare(
                r"SELECT key, value FROM cursorDiskKV WHERE key LIKE 'composerData:%' ESCAPE '\'",
            )?;
            let mut rows = stmt.query([])?;
            while let Some(row) = rows.next()? {
                let key: String = row.get(0)?;
                let id = key.strip_prefix("composerData:").unwrap_or(&key);
                if id.is_empty() {
                    continue;
                }
                let Some(raw) = value_bytes(row.get_ref(1)?) else { continue };
                if let Ok(composer) = serde_json::from_slice::<CursorComposer>(raw) {
                    composers.insert(id.to_string(), composer);
                }
            }
        }

        let mut stmt = conn.prepare(
            r"SELECT key, value FROM cursorDiskKV WHERE key LIKE 'bubbleId:%' ESCAPE '\' ORDER BY key",
        )?;
        let mut rows = stmt.query([])?;
        let mut seen: HashSet<String> = HashSet::new();
        let mut buckets = Tokens::default();
        let mut events = 0;
        // Each bubble maps independently of its siblings, so rows are handled
        // one at a time instead of batched per composer.
        while let Some(row) = rows.next()? {
            let key: String = row.get(0)?;
            let rest = key.strip_prefix("bubbleId:").unwrap_or(&key);
            let Some((composer_id, _)) = rest.split_once(':') else { continue };
            if composer_id.is_empty() {
                continue;
            }
            let Some(composer) = composers.get(composer_id) else { continue };
            let Some(raw) = value_bytes(row.get_ref(1)?) else { continue };
            let Some(event) = map_cursor_bubble(composer_id, composer, &key, raw) else {
                continue;
            };
            if !seen.insert(event.id) {
                continue;
            }
            events += 1;
            buckets.add(event.usage.tokens);
            emit(event.usage);
        }
        Ok(Some((events, buckets)))
    };
    match run() {
        Ok(Some((events, buckets))) => Parsed {
            stats: Some(finish("cursor", started, 1, meta.len(), events, buckets)),
            error: None,
        },
        Ok(None) => Parsed::default(),
        Err(error) => failed(error),
    }
}

#[derive(Debug, Default, Deserialize)]
pub struct CursorComposer {
    #[serde(default, rename = "createdAt")]
    pub created_at: Option<Value>,
    #[serde(default, rename = "modelConfig")]
    pub model_config: Option<Value>,
}

#[derive(Debug, Default, Deserialize)]
pub struct CursorBubble {
    #[serde(default, rename = "bubbleId")]
    pub bubble_id: Option<Value>,
    #[serde(default, rename = "createdAt")]
    pub created_at: Option<Value>,
    #[serde(default, rename = "tokenCount")]
    pub token_count: Option<Value>,
    #[serde(default, rename = "tokenCounts")]
    pub token_counts: Option<Value>,
    #[serde(default, rename = "modelInfo")]
    pub model_info: Option<Value>,
}

#[derive(Debug, PartialEq)]
pub struct CursorEvent {
    pub id: String,
    pub usage: UsageEvent,
}

/// Maps one `bubbleId:<composer>:<bubble>` row to a usage event. No fallback to
/// contextTokensUsed / promptTokenBreakdown: those measure how full the context
/// window is, not tokens consumed.
pub fn map_cursor_bubble(
    composer_id: &str,
    composer: &CursorComposer,
    key: &str,
    raw: &[u8],
) -> Option<CursorEvent> {
    let bubble = serde_json::from_slice::<CursorBubble>(raw).ok()?;
    let counts = as_object(&bubble.token_count).or_else(|| as_object(&bubble.token_counts))?;
    let input = json_int(non_null(counts, "inputTokens").or_else(|| non_null(counts, "input")));
    let output = json_int(non_null(counts, "outputTokens").or_else(|| non_null(counts, "output")));
    if input == 0 && output == 0 {
        return None;
    }
    let mut bubble_id = json_string(bubble.bubble_id.as_ref()).to_string();
    if bubble_id.is_empty() {
        let prefix = format!("bubbleId:{composer_id}:");
        bubble_id = key.strip_prefix(&prefix).unwrap_or(key).to_string();
    }
    let ts = json_time(bubble.created_at.as_ref())
        .or_else(|| json_time(composer.created_at.as_ref()))?;
    let model = cursor_model_name(composer, &bubble);
    Some(CursorEvent {
        id: format!("cursor-bubble-{composer_id}-{bubble_id}"),
        usage: UsageEvent {
            ts,
            agent: "cursor",
            provider: cursor_provider(&model).to_string(),
            model,
            tokens: Tokens {
                input,
                output,
                ..Default::default()
            },
        },
    })
}

pub fn cursor_model_name(composer: &CursorComposer, bubble: &CursorBubble) -> String {
    if let Some(info) = as_object(&bubble.model_info) {
        let name = json_string(info.get("modelName"));
        if !name.is_empty() && name != "default" {
            return name.to_string();
        }
    }
    if let Some(config) = as_object(&composer.model_config) {
        if let Some(first) = config
            .get("selectedModels")
            .and_then(Value::as_array)
            .and_then(|selected| selected.first())
            .and_then(Value::as_object)
        {
            let id = json_string(first.get("modelId"));
            if !id.is_empty() && id != "default" {
                return id.to_string();
            }
        }
        let name = json_string(config.get("modelName"));
        if !name.is_empty() && name != "default" {
            return name.to_string();
        }
    }
    "cursor-auto".to_string()
}

pub fn cursor_provider(model: &str) -> &'static str {
    if model.to_lowercase().contains("grok") {
        "xai"
    } else {
        "cursor"
    }
}

fn as_object(value: &Option<Value>) -> Option<&Map<String, Value>> {
    value.as_ref()?.as_object()
}

/// JSON `null` counts as missing, like a nil interface in Go.
fn non_null<'a>(map: &'a Map<String, Value>, key: &str) -> Option<&'a Value> {
    map.get(key).filter(|value| !value.is_null())
}

pub fn json_string(value: Option<&Value>) -> &str {
    value.and_then(Value::as_str).unwrap_or("")
}

pub fn json_int(value: Option<&Value>) -> i64 {
    match value {
        Some(Value::Number(n)) => n.as_f64().map(|f| f as i64).unwrap_or(0),
        Some(Value::String(s)) => serde_json::from_str::<f64>(s)
            .map(|f| f as i64)
            .unwrap_or(0),
        _ => 0,
    }
}

pub fn json_time(value: Option<&Value>) -> Option<DateTime<Utc>> {
    match value? {
        Value::Number(n) => {
            let f = n.as_f64()?;
            if f > 1e12 {
                Utc.timestamp_millis_opt(f as i64).single()
            } else if f > 1e9 {
                Utc.timestamp_opt(f as i64, 0).single()
            } else {
                None
            }
        }
        Value::String(s) => parse_timestamp(s),
        _ => None,
    }
}
